repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/accesscontrol/AccessControlPolicyPreProcessor.java

```diff
@@ -0,0 +1,161 @@
+import org.apache.atlas.accesscontrol.persona.AtlasPersonaService;
+import org.apache.atlas.accesscontrol.persona.PersonaContext;
+import org.apache.atlas.accesscontrol.purpose.AtlasPurposeService;
+import org.apache.atlas.accesscontrol.purpose.PurposeContext;
+import org.apache.atlas.exception.AtlasBaseException;
+import org.apache.atlas.model.instance.AtlasEntity;
+import org.apache.atlas.model.instance.AtlasEntity.AtlasEntityWithExtInfo;
+import org.apache.atlas.model.instance.AtlasObjectId;
+import org.apache.atlas.model.instance.AtlasStruct;
+import org.apache.atlas.model.instance.EntityMutations;
+import org.apache.atlas.repository.graphdb.AtlasGraph;
+import org.apache.atlas.repository.graphdb.AtlasVertex;
+import org.apache.atlas.repository.store.graph.v2.EntityGraphRetriever;
+import org.apache.atlas.repository.store.graph.v2.EntityMutationContext;
+import org.apache.atlas.repository.store.graph.v2.preprocessor.PreProcessor;
+import org.apache.atlas.type.AtlasTypeRegistry;
+import org.apache.atlas.utils.AtlasPerfMetrics;
+import org.apache.commons.collections.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.atlas.model.instance.EntityMutations.EntityOperation.CREATE;
+import static org.apache.atlas.model.instance.EntityMutations.EntityOperation.DELETE;
+import static org.apache.atlas.model.instance.EntityMutations.EntityOperation.UPDATE;
+import static org.apache.atlas.repository.Constants.POLICY_CATEGORY_PERSONA;
+import static org.apache.atlas.repository.Constants.QUALIFIED_NAME;
+
+public class AccessControlPolicyPreProcessor implements PreProcessor {
+    private static final Logger LOG = LoggerFactory.getLogger(AccessControlPolicyPreProcessor.class);
+
+    private final AtlasTypeRegistry typeRegistry;
+    private final EntityGraphRetriever entityRetriever;
+    private final AtlasPersonaService personaService;
+    private final AtlasPurposeService purposeService;
+
+    public AccessControlPolicyPreProcessor(AtlasTypeRegistry typeRegistry, AtlasGraph graph, EntityGraphRetriever entityRetriever) {
+        this.entityRetriever = entityRetriever;
+        this.typeRegistry = typeRegistry;
+        personaService = new AtlasPersonaService(graph, entityRetriever);
+        purposeService = new AtlasPurposeService(graph, entityRetriever);
+    }
+
+    @Override
+    public void processAttributes(AtlasStruct entityStruct, EntityMutationContext context, EntityMutations.EntityOperation operation) throws AtlasBaseException {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("AccessControlPolicyPreProcessor.processAttributes: pre processing {}, {}",
+                    entityStruct.getAttribute(QUALIFIED_NAME), operation);
+        }
+
+        AtlasEntity entity = (AtlasEntity) entityStruct;
+        AtlasVertex vertex = context.getVertex(entity.getGuid());
+
+        switch (operation) {
```

Comment:
what about other operations like PARTIAL_UPDATE, DELETE, PURGE? Should we add a default case?

Reply:
There is no special handling in the case of PARTIAL_UPDATE

PURGE is not needed to handle as it PURGE required entity to be in the deleted state already & DELETE is handled in AtlasEntityStoreV2.deleteVertices(..)

Reply:
for PARTIAL_UPDATE, explicitly passing UPDATE